ekncontent/ekncontent/eknc-domain.c

```diff
@@ -0,0 +1,862 @@
+      return FALSE;
+
+  if (!g_file_query_exists (manifest_file, cancellable))
+    {
+      g_autoptr(GFile) bundle_manifest_file = g_file_get_child (bundle_dir, "manifest.json");
+      if (!g_file_query_exists (bundle_manifest_file, cancellable))
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                       "You have no manifest.json and are not running from a "
+                       "Flatpak bundle. You must download a content update");
+          return FALSE;
+        }
+      if (!g_file_make_symbolic_link (manifest_file,
+                                      g_file_get_path (bundle_manifest_file),
+                                      cancellable,
+                                      error))
+        return FALSE;
+    }
+
+
+  g_autofree gchar *contents = NULL;
+  if (!g_file_load_contents (manifest_file, cancellable, &contents,
+                             NULL, NULL, error))
+    return FALSE;
+
+  g_autoptr(JsonNode) manifest_node = NULL;
+  if (!(manifest_node = json_from_string (contents, error)))
+    return FALSE;
+
+  if (!JSON_NODE_HOLDS_OBJECT (manifest_node))
+    {
+      g_set_error (error, EKNC_DOMAIN_ERROR, EKNC_DOMAIN_ERROR_BAD_MANIFEST,
+                   "Manifest file does not hold a json object");
+      return FALSE;
+    }
+
+  JsonNode *shards_node = json_object_get_member (json_node_get_object (manifest_node), "shards");
+
+  if (shards_node == NULL || !JSON_NODE_HOLDS_ARRAY (shards_node))
+    {
+      g_set_error (error, EKNC_DOMAIN_ERROR, EKNC_DOMAIN_ERROR_BAD_MANIFEST,
+                   "Missing or malformed shards entry in manifest");
+      return FALSE;
+    }
+
+  if (!eknc_domain_clean_old_symlinks (self, cancellable, error))
+    return FALSE;
+
+  JsonArray *shards_array = json_node_get_array (shards_node);
+  for (guint i = 0; i < json_array_get_length (shards_array); i++)
+    {
+      JsonNode *shard_node = json_array_get_element (shards_array, i);
+      if (!JSON_NODE_HOLDS_OBJECT (shard_node))
+        {
+          g_set_error (error, EKNC_DOMAIN_ERROR, EKNC_DOMAIN_ERROR_BAD_MANIFEST,
+                       "Malformed shard entry in manifest");
+          return FALSE;
+        }
+      JsonNode *path_node = json_object_get_member (json_node_get_object (shard_node), "path");
+      if (path_node == NULL || json_node_get_value_type (path_node) != G_TYPE_STRING)
+        {
+          g_set_error (error, EKNC_DOMAIN_ERROR, EKNC_DOMAIN_ERROR_BAD_MANIFEST,
+                       "Malformed shard entry in manifest");
+          return FALSE;
+        }
+      g_autoptr(GFile) bundle_shard_file = g_file_get_child (bundle_dir, json_node_get_string (path_node));
+      g_autoptr(GFile) subscription_shard_file = g_file_get_child (subscription_dir, json_node_get_string (path_node));
+
+      // Make a symlink if necessary
+      if (g_file_query_exists (bundle_shard_file, cancellable) &&
```

Comment:
Well a lot of disk io happening here, and I think the double check was already there. Probably not the biggest efficiency win. But I'll try writing out the other way.